plugins/bufferlines/barbar.nix

```diff
@@ -69,249 +24,567 @@ let
+        default = { };
+        description = helpers.defaultNullOpts.mkDesc {
+          "vim.diagnostic.severity.ERROR" = {
+            enabled = false;
+            icon = " ";
+          };
+          "vim.diagnostic.severity.HINT" = {
+            enabled = false;
+            icon = "󰌶 ";
+          };
+          "vim.diagnostic.severity.INFO" = {
+            enabled = false;
+            icon = " ";
+          };
+          "vim.diagnostic.severity.WARN" = {
+            enabled = false;
+            icon = " ";
+          };
+        } "Set the icon for each diagnostic level.";
+      };
+
+      gitsigns =
+        helpers.defaultNullOpts.mkAttrsOf
+          (
+            with types;
+            submodule {
+              options = {
+                enabled = mkOption {
+                  type = types.bool;
+                  description = helpers.defaultNullOpts.mkDesc true ''
+                    Enables showing git changes of this type.
+                    Requires `|gitsigns.nvim|`.
+                  '';
+                };
+
+                icon = mkOption {
+                  type = types.str;
+                  example = "";
+                  description = ''
+                    The icon which accompanies the number of git status.
+
+                    To disable the icon but still show the count, set to an empty string.
+                  '';
+                };
+              };
+            }
+          )
+          {
+            added = {
+              enabled = true;
+              icon = "+";
+            };
+            changed = {
+              enabled = true;
+              icon = "~";
+            };
+            deleted = {
+              enabled = true;
+              icon = "-";
+            };
+          }
+          "Gitsigns icons.";
+
+      filename = helpers.defaultNullOpts.mkBool true ''
+        If `true`, show the name of the file.
+      '';
+
+      filetype = {
+        custom_colors = helpers.defaultNullOpts.mkBool false ''
+          If `true`, the `Buffer<status>Icon` color will be used for icon colors.
+        '';
+
+        enabled = helpers.defaultNullOpts.mkBool true ''
+          Filetype `true`, show the `devicons` for the associated buffer's `filetype`.
+        '';
+      };
+
+      separator = {
+        left = helpers.defaultNullOpts.mkStr "▎" ''
+          The left separator between buffers in the tabline.
+        '';
+
+        right = helpers.defaultNullOpts.mkStr "" ''
+          The right separator between buffers in the tabline.
+        '';
+
+        separator_at_end = helpers.defaultNullOpts.mkBool true ''
+          If true, add an additional separator at the end of the buffer list.
+          Can be used to create a visual separation when the inactive buffer background color is the
+          same as the fill region background color.
+        '';
+      };
+
+      modified = helpers.defaultNullOpts.mkAttrsOf types.anything { button = "●"; } ''
```

Comment:
Why `anything` and not `str`?

Reply:
Because this can litterally be any of the surrounding options.
The way to declare this cleanly would be to express this strange recursivity. That is what was done prior to this rewrite.

Reply:
Ah ok, IMO this is fine then

Reply:
Maybe a comment would be useful for future reference though

Reply:
https://github.com/nix-community/nixvim/pull/1608#discussion_r1623296130

Reply:
Could you add a comment that you're using `anything` to avoid needing to implement recursive options?